### ISSUE TITLE
Add support for moss tts 1.5

### ISSUE THREAD

**Issue by ILG2021** [0xShug0/audio.cpp#607]
High-quality, fast model based on the delay paradigm, although 8b.

**Comment by 0xShug0** [comment]
Hi @ILG2021 We have MOSS-TTS-Local-Transformer-v1.5. Is that the one you were thinking of?

Also @jrohde may be working on MOSS-TTSD https://github.com/0xShug0/audio.cpp/pull/278#issuecomment-5395502583

**Comment by ILG2021** [comment]
local transformer 1.5 is local pattern, not the delay pattern. you can refer to:
https://github.com/pwilkin/openmoss

**Comment by 0xShug0** [comment]
Added to waiting list!

**Comment by christopherthompson81** [comment]
Before I start on this — @0xShug0 @jrohde — the model here and MOSS-TTSD are the same architecture, which I don't think has been said out loud yet, and it changes who should do what.

Fetching both configs directly:

```
OpenMOSS-Team/MOSS-TTS-v1.5    architectures ['MossTTSDelayModel']  model_type moss_tts_delay  n_vq 32
OpenMOSS-Team/MOSS-TTSD-v1.0   architectures ['MossTTSDelayModel']  model_type moss_tts_delay  n_vq 16
```

Same `architectures`, same `model_type`, same `audio_vocab_size` (1024), same special tokens — `audio_start_token_id` 151652, `audio_end_token_id` 151653, `audio_assistant_delay_slot_token_id` 151662, and the rest. **They differ in `n_vq`: 32 against 16.**

So this isn't "adjacent to" the MOSS-TTSD work @jrohde offered on [#278](https://github.com/0xShug0/audio.cpp/pull/278#issuecomment-5395502583) and you approved — it's one config value away from it. jrohde's own description there ("`MossTTSDelay`, n_vq 16, the same special tokens and the same 24 kHz v1 codec") describes v1.5 exactly apart from the codebook count. Whoever ports one gets the other very nearly free, and if two of us start independently we'll write the same ~4000 lines twice.

I'd rather ask than find that out afterwards, so: **@jrohde, have you started on TTSD?** If you have, this should be yours — the delta is a config value and whatever differs in the prompt format, and I'd be glad to take a different issue. If you haven't, or you'd rather not, I'm happy to do `moss_tts_delay` as a family that covers both `n_vq` values and hand TTSD's prompt/multi-speaker side back to you.

Two smaller things worth recording while I was checking:

**@ILG2021 is right, and the distinction is real.** `moss_tts_local` (already in the tree) is `OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5` — Qwen3-4B, tokenizer v2, 48 kHz stereo, 12 codebooks through a depth transformer. This issue is `OpenMOSS-Team/MOSS-TTS-v1.5` — Qwen3-8B, tokenizer v1, 24 kHz mono, 32 codebooks on a delay pattern. Different model, not a version bump.

**A good chunk already exists.** #278 landed `src/community_models/moss_voicegen`, whose `delay_decoder.cpp` is already driven by `config_.num_codebooks` rather than a fixed count, along with the v1 codec path, the text tokenizer and the heads. Either 8B port starts from there rather than from scratch.

One thing I'd like to do either way, and would share whoever ends up writing the port: run the capability evaluation against the upstream PyTorch *before* any C++ exists — what actually works, at what durations, in which languages. On the last model I worked on here that evaluation needed no C++ at all, would have been decisive on day one, and instead ran last. Happy to post the results on this issue regardless of who takes the port.

**Comment by christopherthompson81** [comment]
Picking this up. Some findings that change the shape of the work, and one framework bug fixed along the way — branch `moss-tts-delay` on my fork, one commit so far.

## The overlap with MOSS-TTSD is bigger than I said above

I claimed the two models differ by one config value. Having now read both reference implementations, that undersold it. Diffing them:

```
inference_utils.py          0 differing lines
configuration_moss_tts.py   4
modeling_moss_tts.py       84     <- sampling defaults + a generation_config fallback
processing_moss_tts.py     99     <- the prompt
```

Diffing **only the delay generation loop**, comments stripped, the entire difference is:

```
TTSD:  audio_ch0_logits[..., 1024] = float('-inf')
v1.5:  audio_ch0_logits[..., self.config.audio_pad_code] = float('-inf')
```

and `audio_pad_code` is 1024 in both. The loop is functionally identical. The prompts differ in enumerable ways — TTSD has `- Scene:`, v1.5 has `- Tokens:` and a text normaliser, TTSD emits `[S{i}]: None` for absent speakers where v1.5 omits the slot. Both use the same `[S{i}]:\n<audio>` multi-speaker convention, so that part is shared rather than TTSD-only.

**@jrohde** — TTSD is still yours as far as I'm concerned. I'm writing the family `n_vq`-generic because that is what the existing code already is, so if you want TTSD the remaining work should be a spec, the `scene` field and the None-branch rather than a port. Happy to hand over whatever is useful, or to stay out of it entirely if you're mid-flight.

## A framework bug: v1 codec encode has never run

`moss_tts_local` exercises **v2** encode and `moss_voicegen` exercises **v1** decode, but nothing calls **v1 encode** — so its `encoder_stages` had never executed. The delay models clone from a reference recording, which is what calls it.

It fails immediately: `MOSS codec tensor not found: encoder.3.input_proj.weight`. The checkpoint's projections are asymmetric, because upstream only creates a projection when a stage changes width:

```
encoder.1: input_proj, output_proj     decoder.0: input_proj
encoder.3: output_proj                 decoder.2: input_proj
encoder.5: output_proj                 decoder.4: input_proj
encoder.7: output_proj                 decoder.6: input_proj, output_proj
```

The loader already knew this for `output_proj` — there's a comment saying v1 omits it wherever `output_dimension == d_model` — but loaded `input_proj` unconditionally. The same width rule predicts the checkpoint exactly, so the config was right and only the loader was wrong. It survived because v1's *decoder* widens on input at every stage, so handling the output side alone satisfies the decoder completely and only the encoder trips.

Fixed in shared code, so v2 and nano inherit the rule; it only changes behaviour where the old code threw.

## Parity, both paths

Neither test needs a GGUF conversion — the checkpoints' tensor names are already what the runtimes look up, so both read the HF download through `open_tensor_source()`.

| test | result |
|---|---|
| `moss_voicegen_codec_encode_parity` (v1 encode) | **3072/3072 codes exact** |
| `moss_tts_delay_backbone_parity` (8B, n_vq 32) | prefill max_rel **0.0052**, cached single-step **0.0038**, `non_finite=0`, passes at tolerance 0.006 |

Worth knowing for anyone running these: **the RLFQ quantizer is sensitive enough that the reference disagrees with itself across devices.** A CUDA reference and a CPU reference differ in 36 of 3072 codes, at frames 63/64/75/78. That first read as a boundary bug in my port. Exact code match gates the harness being device-matched as much as the encoder.

## Capability check on the reference, before the port

Run against upstream PyTorch, transcribed with Whisper, pitch-tracked with `pyin`:

| case | audio | RTF | median F0 | transcript |
|---|---|---|---|---|
| reference recording | 7.73 s | — | 189.4 Hz | (voice being cloned) |
| instruction-only | 4.16 s | 1.05 | 90.9 Hz | exact |
| **voice clone** | 4.64 s | 1.04 | **188.3 Hz** | exact |
| Chinese | 3.76 s | 0.78 | 122.1 Hz | exact |
| long form | 20.40 s | 0.53 | 135.5 Hz | correct |
| `tokens=40` | 3.28 s | 0.84 | 91.5 Hz | exact |

Cloning tracks the reference closely, Chinese works from an English instruction, long form works and RTF *improves* with length, and `tokens` is a real, honoured duration budget (40 tokens → 41 frames at 12.5 Hz).

**One limitation worth recording now rather than after merge:** voice-attribute instructions are not reliably followed. Four prompts, same seed, differing only in requested speaker:

```
"A high-pitched young woman's voice, clearly female."   -> 216.3 Hz   plausible
"A woman speaking softly and warmly."                   -> 184.5 Hz   borderline
"A man speaking in a low register."                     -> 187.8 Hz   miss
"A very deep, low-pitched man's voice."                 -> 180.8 Hz   clear miss
```

Caveats: five samples, one seed each, and median F0 is a proxy — perceived gender is not pitch alone. It supports "unreliable", not "never works". But a "very deep, low-pitched man's voice" at 181 Hz is wrong by any measure, and cloning is clearly the capability that works here. If this ships I'd suggest a limitations line saying so, since voice *design* from a written instruction is what `moss_voicegen` is already for.

## What's left

The pipeline is mostly present: the delay state machine is already `n_vq`-generic, the backbone and heads are config-driven, the tokenizer and both codec directions are done. What remains is the v1.5 prompt builder, a port of `normalize_tts_text` (346 lines of `re`/`unicodedata`, and it ships its own test vectors), the model spec, GGUF conversion and docs.

One question before I go further, since it touches merged code: **#332 promoted the MOSS codec runtime into the framework so the family could share one runtime.** The same argument applies to the delay decoder, the backbone wrapper and the heads — about 1400 of `moss_voicegen`'s 2167 lines are family-generic and would let v1.5 and TTSD sit on top instead of duplicating them. Would you rather I do that promotion, or keep v1.5 self-contained and accept the duplication? Happy either way; it's your blast radius, not mine.

**Comment by 0xShug0** [comment]
@christopherthompson81 Thanks! I will review your moss tts and kokoro PRs tomorrow. Promoting the delay arch as a shared runtime makes sense. Please ensure that the runtime remains configurable and is responsible for lifecycle management, state management, and core logic, while model-specific logic stays local to each adopter. 

Udpate: BTW, the MOSS family seems relatively easy to port since the models reuse many existing building blocks. moss_transcribe_diarize  only took about 500 lines of C++ code.

**Comment by christopherthompson81** [comment]
GGUF packages are built, verified and published. Branch `moss-tts-delay` on my fork now has the whole port.

## Packages

**https://huggingface.co/christopherthompson81/MOSS-TTS-v1.5-GGUF**

| file | backbone | codec | size | transcript accuracy |
|---|---|---|---|---|
| `moss_tts_v15_bf16_codec_f16.gguf` | bf16 | f16 | 20.5 GB | exact |
| `moss_tts_v15_q8_0_codec_f16.gguf` | q8_0 | f16 | 13.2 GB | 4/4 takes exact, WER 0.000 |
| `moss_tts_v15_q4_k_codec_f16.gguf` | q4_k | f16 | 9.3 GB | 3/4 takes exact, WER 0.021 |

Each is self-contained: backbone, codec (**encoder and decoder** — cloning needs the encoder, unlike `moss_voicegen`'s decode-only package), tokenizer and model spec. The codec is held at f16 in every package; it is a small share of the total and it is what turns codes back into a waveform, so quantising it buys little and risks the part you actually hear.

**@0xShug0 — these are in my namespace because the spec's `package_defaults.download.repo` is `audio-cpp/audio.cpp-gguf`, which I can't write to.** Happy for you to mirror them; I left the spec pointing at the canonical location rather than hardcoding my namespace into the project, which means the download path is correct once mirrored and broken until then. Say the word if you'd rather I point it at mine in the meantime.

## ⚠ q4_k is not lossless on the words

Worth stating plainly, because it would be easy to ship as equivalent. All three packages load, run, and clone the reference equally well — median F0 180.8, 194.4 and 185.1 Hz against the reference recording's 189.4. On any check that asks "did it produce audio", q4_k passes cleanly.

Transcribing it did not:

```
q8_0   4/4 takes exact,  mean WER 0.000
q4_k   3/4 takes exact,  mean WER 0.021   ("The sentence" for "This sentence")
```

Both `whisper-base.en` and `whisper-small` agree on the substitution, so it is the audio rather than a transcription artifact — worth confirming, since a small ASR disagreeing on an article is usually the ASR.

Four takes of one sentence does not pin the true rate down, and the model card says so. But it is enough that q4_k should not be described as equivalent to q8_0. The voice survived intact in every take; it was the wording that moved — which is the part a waveform check never looks at.

## The branch

```
1f01257c  codec: load a stage's input projection only when it exists
00fef1b4  promote the delay-family runtimes into the framework
511c698a  the prompt builder, with reference audio
5093c2f7  the model, end to end
dbaddc9e  GGUF packaging and docs
d950408a  name the loader factory the way the catalog check expects
4929779f  port the MOSS-TTS input normaliser
```

Full build clean, ctest 68/68, loader/catalog sync passes. Parity: v1 codec encode 3072/3072 exact; backbone 0.0052 prefill and 0.0038 cached-step at n_vq 32; prompt parity on both the instruction-only and cloning shapes with zero text and zero code mismatches; text normalisation 38/38 on the vectors that ship inside the reference script. `moss_voicegen`'s own five suites are unchanged after the promotion — I downloaded its checkpoint and ran them rather than inferring from the v1.5 path.

Ready for review whenever you have time, and still happy to hand the delay-family groundwork to @jrohde for TTSD — the family runtimes are `n_vq`-generic, so that should now be a spec, the `scene` prompt field and the None-branch rather than a port.

**Comment by jrohde** [comment]
@christopherthompson81 Thanks for checking. I haven't really started on TTSD yet, and realistically I won't get to it in the coming weeks either. So it would be great if you took on TTSD along with the rest of the `moss_tts_delay` family. You're clearly well ahead, and one person building the whole family indeed makes more sense than splitting it.

If it helps, I'm happy to test on AMD and Intel hardware once there's something to run.

**Comment by ILG2021** [comment]
@0xShug0 I have used openmoss cpp, so this feature is not needed. thanks.

**Comment by 0xShug0** [comment]
If it’s an audio model, it'll probably end up in audio.cpp sooner or later :)

@christopherthompson81 I'm fine with pointing to your repo in the specs. 

@jrohde Thanks for weighing in!
